File: TicTacToe/TicTacToe.py
import random

class Board:
    WIN = (
        ((0, 0), (0, 1), (0, 2)), ((1, 0), (1, 1), (1, 2)),
        ((2, 0), (2, 1), (2, 2)), ((0, 0), (1, 0), (2, 0)),
        ((0, 1), (1, 1), (2, 1)), ((0, 2), (1, 2), (2, 2)),
        ((0, 0), (1, 1), (2, 2)), ((2, 0), (1, 1), (0, 2)),
    )
    def __init__(self):
        self.board = [[" " for _ in range(3)] for _ in range(3)]

    def __str__(self):
        return "\n" + "\n".join([" | ".join(x) for x in self.board]) + "\n"

    def is_valid(self, pos):
        return self.get_item(pos) == " "

    def all_moves(self):
        yield from ((r, c) for r in range(3) for c in range(3)
                    if self.is_valid((r, c)))

    def state(self):
        for p in Board.WIN:
            for i in ("X", "O"):
                if all((self.get_item(x) == i for x in p)): return i
        try:
            next(self.all_moves())
            return "ongoing"
        except StopIteration: return "draw"
        
    def get_item(self, pos):
        return self.board[pos[0]][pos[1]]

    def set_item(self, pos, piece):
        self.board[pos[0]][pos[1]] = piece


def pick_next(board, player, count=0):
    def pick_next_help(board, player, count):
        other = "X" if player == "O" else "O"
        s = board.state()
        if s == player: return (), 10, count
        elif s == other: return (), -10, count
        count += 1
        best = False
        for (r, c) in board.all_moves():
            board.set_item((r, c), player)
            _, s, co = pick_next_help(board, other, count)
            s *= -1

            if (not best or s > best[1] or (s == best[1] and (
                (s >= 0 and co < best[2]) or (s < 0 and co > best[2])
            ))): best = (r, c), s, co

            board.set_item((r, c), " ")

        if best: return best
        else: return (), 0, count
    return pick_next_help(board, player, count)[0]


def play(first=True):
    board, i = Board(), 0
    if not first:
        board.set_item(random.choice([(0, 0), (0, 2), (2, 0), (2, 2)]), "X")
    first = "X" if first else "O"
    other = "X" if first == "O" else "O"
    while board.state() == "ongoing":
        i += 1
        print(board)
        if i % 2:
            s = input("Row, Col: ")
            r, c = int(s[0]), int(s[1])
            if board.is_valid((r, c)): board.set_item((r, c), first)
            else: i -= 1
        else: board.set_item(pick_next(board, other), other)
    print(board)
    print(board.state())


if __name__ == '__main__':
    import sys
    play(not bool(sys.argv[1:]))

File: TicTacToe/tictactoe.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>


const int WIN[8][3][2] = {
    {{0, 0}, {0, 1}, {0, 2}}, 
    {{1, 0}, {1, 1}, {1, 2}},
    {{2, 0}, {2, 1}, {2, 2}}, 
    {{0, 0}, {1, 0}, {2, 0}},
    {{0, 1}, {1, 1}, {2, 1}}, 
    {{0, 2}, {1, 2}, {2, 2}},
    {{0, 0}, {1, 1}, {2, 2}}, 
    {{2, 0}, {1, 1}, {0, 2}}
};


typedef int Board[3][3];
typedef int ValidMoves[9][2];


void initialize_board(Board board) {
    for(int r=0; r<3; r++) {
        for (int c=0; c<3; c++) {
            board[r][c] = 0;
        }
    }
}

void Board_print(Board board) {
    char a[3];
    for (int r=0; r<=2; r++) {
        for(int i=0; i<3; i++) {
            switch(board[r][i]) {
            case 0:
                a[i] = ' ';
                break;
            case 1:
                a[i] = 'X';
                break;
            case 2:
                a[i] = 'O';
                break;
            }
        }

        printf("%c | %c | %c \n", a[0], a[1], a[2]);
        if(r < 2)
            printf("-----------\n");
    }
}

inline bool Board_is_valid(Board board, int row, int col) {
    if(row < 0 || row > 2 || col < 0 || col > 2) 
        return false;
    return board[row][col] == 0;
}

void Board_all_moves(Board board, ValidMoves valid_moves) {
    int index = 0;
    for(int r=0; r<=2; r++) {
        for (int c=0; c<=2; c++) {
            if(Board_is_valid(board, r, c)) {
                valid_moves[index][0] = r;
                valid_moves[index][1] = c;
                index++;
            }
        }
    }
    if(index < 9) {
        valid_moves[index][0] = -1;
    }
}

bool all(Board board, int win, int player) {
    // return true if all positions in WIN[win]
    // are occupied by the player
    for(int j=0; j<3; j++) {
        if(board[WIN[win][j][0]][WIN[win][j][1]] != player)
            return false;
    }
    return true;
}


int Board_state(Board board) {
    for(int win=0; win<8; win++) {
        if(all(board, win, 1)) {
            return 1;
        } else if(all(board, win, 2)) {
            return 2;
        }
    }
    for(int r=0; r<3; r++) {
        for(int c=0; c<3; c++) {
            if(Board_is_valid(board, r, c))
                return 0;
        }
    }

    return 3;
}


typedef struct Status {
    int row, col, score, count;
} Status;

Status pick_next(Board board, int player, int count) {
    Status s = {0};
    s.count = count;
    int other = (player == 2) ? 1 : 2;
    int state = Board_state(board);
    if(state == player) {
        s.score = 10;
        return s;
    } else if (state == other) {
        s.score = -10;
        return s;
    }
    count++;

    int row, col;

    Status best = {-1};
    ValidMoves valid_moves; 
    Board_all_moves(board, valid_moves);
    for(int move=0; move<9; move++) {
        if(valid_moves[move][0] == -1) {
            break;
        }
        row = valid_moves[move][0]; col = valid_moves[move][1];
        board[row][col] = player;
        s = pick_next(board, other, count);
        board[row][col] = 0;
        s.score *= -1;
        if(best.row == -1 || s.score > best.score || (s.score == best.score && ((s.score >= 0 && s.count < best.count) || (s.score < 0 && s.count > best.count)))) {
            best.row = row; best.col = col; best.score = s.score; best.count = s.count;
        }
    }

    if(best.row == -1) {
        best.row = -1; best.col = -1; best.score = 0; best.count = count;
    }
    return best;
}


void play(bool first) {
    Board board;
    initialize_board(board);
    int i = 0;
    if(!first) {
        // board[0][0] = 1;
        i--;
    }

    int r, c;

    int first_player = first ? 1 : 2;
    int other_player = first_player == 2 ? 1 : 2;

    while(Board_state(board) == 0) {
        i++;
        Board_print(board);
        if(i % 2) {
            printf("Row, col:\n");
            scanf("%d,%d", &r, &c);
            if(Board_is_valid(board, r, c)) {
                board[r][c] = first_player;
            } else {
                printf("Invalid move\n");
                i--;
            }
        } else {
            Status status = pick_next(board, other_player, 0);
            board[status.row][status.col] = other_player;
        }
    }

    Board_print(board);
    switch(Board_state(board)) {
    case 1:
        printf("Player X wins\n");
        break;
    case 2:
        printf("Player O wins\n");
        break;
    case 3:
        printf("It's a tie\n");
        break;
    }
}


int main() {
    int first;
    printf("Play first?\n");
    scanf("%d", &first);
    play(first);
    return 0;
}